fpga_interchange/arch_pack_clusters.cc

```diff
@@ -282,6 +276,148 @@ bool Arch::getClusterPlacement(ClusterId cluster, BelId root_bel,
     return true;
 }
 
+static dict<int32_t, dict<IdString, BelId>> tileAndBelNameToBelIdCache;
+
+BelId check_and_return(int32_t tile, IdString name){
+    if(tileAndBelNameToBelIdCache.count(tile)
+       && tileAndBelNameToBelIdCache[tile].count(name))
+        return tileAndBelNameToBelIdCache[tile][name];
+    else
+        return BelId();
+}
+
+void add_to_cache(int32_t tile, IdString name, BelId t){
+    tileAndBelNameToBelIdCache[tile][name] = t;
+}
+
+bool find_site_idx(const Context *ctx, const ClusterPOD &cluster, BelId root_bel, uint32_t &idx){
+    bool found = false;
+    const auto &site_inst = ctx->get_site_inst(root_bel);
+    IdString site_type(site_inst.site_type);
+
+    if (ctx->debug){
+        log_info("%s\n", ctx->get_site_name(root_bel));
+        log_info("Root_bel site_type: %s\n", site_type.c_str(ctx));
+        log_info("Allowed site_types:\n");
+    }
+    for(const auto &site : cluster.physical_placements){
+        IdString name(site.site_type);
+        if(ctx->debug)
+            log_info("\t%s\n", name.c_str(ctx));
+
+        if (name == site_type){
+            found = true;
+            break;
+        }
+        idx++;
+    }
+    return found;
+}
+
+bool find_placement_idx(const Context *ctx, const ClusterPOD &cluster,
+                        BelId root_bel, uint32_t idx, uint32_t &placement_idx){
+    bool found = false;
+    const auto &bel_data = bel_info(ctx->chip_info, root_bel);
+    IdString root_bel_name(bel_data.name);
+    if(ctx->debug){
+        log_info("Root_bel name: %s\n", root_bel_name.c_str(ctx));
+        log_info("Allowed root_bels:\n");
+    }
+    for(const auto &place : cluster.physical_placements[idx].places){
+        for (const auto bel : place.bels){
+            IdString name(bel);
+            if(ctx->debug)
+                log_info("\t%s\n",name.c_str(ctx));
+
+            if(name == root_bel_name){
+                found = true;
+                break;
+            }
+        }
+        if (found)
+            break;
+        placement_idx++;
+    }
+    return found;
+}
+
+dict<uint32_t, BelId> idx_bel_mapping(const Context *ctx, BelId root_bel,
+                                      const ClusterPOD &cluster, uint32_t idx, uint32_t placement_idx){
+    dict<uint32_t, BelId> idx_bel_map;
+    auto root_bel_full_name = ctx->getBelName(root_bel);
+    uint32_t t_idx = 0;
+    if(ctx->debug)
+        log_info("Used bels:\n");
+    for(const auto &bel : cluster.physical_placements[idx].places[placement_idx].bels){
+        IdString s_bel(bel);
+        BelId t = check_and_return(root_bel.tile, s_bel);
+        IdStringList cpy(root_bel_full_name.size());
+        if (t == BelId()){
+            for(uint32_t j = 0; j < root_bel_full_name.size(); j++)
+                cpy.ids[j] = root_bel_full_name[j];
+            cpy.ids[1] = s_bel;
```

Comment:
index `size()-1` is probably better than index 1 if a name had more than 2 components ever

Reply:
Fixed in next commit